### PR TITLE
Added Exceptions#1 Introduction to Error Codes for Exceptions in the Application

### PR DESCRIPTION
**Pull request by prolevel1**
### When trying to deal with exception handling, in the response we should get something like below

```
{
   “Error” : {
                     “ErrorCode : 10001”,
                     ”errorDescription : “OBJECT_NOT_FOUND”,
                     ”debugCode” : 1890
                   }
}
```

### CODE REVIEW

File: src/main/java/com/practice/PRACTICE/controller/EmployeeController.java [hunk 1]
```diff
@@ -38,4 +47,8 @@ public ResponseEntity<?> deleteById(@PathVariable long id) throws ResourceNotFou
     public ResponseEntity<Employee> updateEmp(@RequestBody Employee employee, @PathVariable long id) throws ResourceNotFoundException {
         return  ResponseEntity.ok(empService.updateEmp(employee, id));
     }
+    @GetMapping("/get/{fname}/{lname}")
```
Comment (Shubham-Xperi) [inline src/main/java/com/practice/PRACTICE/controller/EmployeeController.java]: change `@GetMapping` to `@PostMapping`

File: src/main/java/com/practice/PRACTICE/service/EmpServiceImpl.java [hunk 2]
```diff
@@ -44,6 +45,22 @@ public Employee updateEmp(Employee employee, long id) throws ResourceNotFoundExc
...
+        Employee res = null;
+        for (Employee e : employees) {
+            if (e.getLastName().equalsIgnoreCase(lName) && e.getFirstName().equalsIgnoreCase(fName)) {
+                res =   e;
+                break;
+            }
+
+        }
+        if (res == null) {
+            throw  new ObjectNotFound("OBJECT_NOT_FOUND");
```
Comment (Shubham-Xperi) [inline src/main/java/com/practice/PRACTICE/service/EmpServiceImpl.java]: create a constant variable for this `OBJECT_NOT_FOUND` string. Better to create a `static class` inside `util` package that will have constants for everything which can be used throughout the project. below is an example

```
class Constants {
    private final String OBJECT_NOT_FOUND = "object not found" ;
    private final String INVALID_INPUT_PARAMETERS  = "input parameters are not valid in the request body";
    etc ...
}
```

File: src/test/java/com/practice/PRACTICE/service/EmpServiceImplTest.java [hunk 3]
```diff
@@ -35,17 +35,17 @@ class EmpServiceImplTest {
...
     @Test
     public void getAllEmpTest() {
         Employee employee1 = Employee.builder()
                 .empId(2)
-                .firstName("Raj")
-                .lastName("Kumar")
+                .firstName("ahs")
+                .lastName("pqrs")
                 .build();
         given(empRepo.findAll()).willReturn(List.of(employee, employee1));
```
Comment (Shubham-Xperi) [inline src/test/java/com/practice/PRACTICE/service/EmpServiceImplTest.java:51]: Please add Unit test cases for negative scenarios as well, where exceptions can occur.

File: src/main/java/com/practice/PRACTICE/controller/EmployeeController.java [hunk 4]
```diff
@@ -1,12 +1,21 @@
 package com.practice.PRACTICE.controller;
 
 import com.practice.PRACTICE.entity.Employee;
```
Comment (Shubham-Xperi) [inline src/main/java/com/practice/PRACTICE/controller/EmployeeController.java]: All the import statements should be in **alphabetical** order.

File: src/main/java/com/practice/PRACTICE/service/EmpServiceImpl.java [hunk 5]
```diff
@@ -13,6 +14,8 @@
 public class EmpServiceImpl implements EmpService{
     @Autowired
     private EmpRepo empRepo;
+    @Autowired
+    private Constants constants;
```
Comment (Shubham-Xperi) [inline src/main/java/com/practice/PRACTICE/service/EmpServiceImpl.java]: The Class constant should be static so that you do not need to create any bean or dependency to the other classes.

File: src/main/java/com/practice/PRACTICE/service/EmpServiceImpl.java [hunk 6]
```diff
@@ -57,7 +61,8 @@ public Employee findByName(String fName, String lName) throws ObjectNotFound {
 
         }
         if (res == null) {
-            throw  new ObjectNotFound("OBJECT_NOT_FOUND");
+           // throw  new ObjectNotFound("OBJECT_NOT_FOUND");
+            throw new ObjectNotFound(constants.getOBJECT_NOT_FOUND());
```
Comment (Shubham-Xperi) [inline src/main/java/com/practice/PRACTICE/service/EmpServiceImpl.java]: it should be like `Constants.OBJECT_NOT_FOUND`

File: src/main/java/com/practice/PRACTICE/service/EmpServiceImpl.java [hunk 7]
```diff
@@ -61,8 +60,8 @@ public Employee findByName(String fName, String lName) throws ObjectNotFound {
 
         }
         if (res == null) {
-           // throw  new ObjectNotFound("OBJECT_NOT_FOUND");
-            throw new ObjectNotFound(constants.getOBJECT_NOT_FOUND());
+
+            throw new ObjectNotFound(OBJECT_NOT_FOUND);
```
Comment (Shubham-Xperi) [inline src/main/java/com/practice/PRACTICE/service/EmpServiceImpl.java]: it should be like `Constants.OBJECT_NOT_FOUND`
if used like this then this named variable can be used through out the project which is not a good practice.

File: src/main/java/com/practice/PRACTICE/util/Constants.java [hunk 8]
```diff
@@ -1,10 +1,12 @@
...
 
 import lombok.Getter;
-import org.springframework.stereotype.Component;
 
-@Component
-@Getter
-public class Constants {
-    private final String OBJECT_NOT_FOUND = "object not found" ;
+
+public  class Constants {
```
Comment (Shubham-Xperi) [inline src/main/java/com/practice/PRACTICE/util/Constants.java]: make this class final

File: src/main/java/com/practice/PRACTICE/util/Constants.java [hunk 9]
```diff
@@ -1,10 +1,12 @@
...
+    private Constants() {
```
Comment (Shubham-Xperi) [inline src/main/java/com/practice/PRACTICE/util/Constants.java]: Remove this constructor

**Comment by Shubham-Xperi** [comment]
Looks good to me. great!
